mlflow/fastai.py

```diff
@@ -295,63 +300,64 @@ def autolog(log_models=True):
-                try_mlflow_log(mlflow.log_param, "train_bn", self.train_bn)
-
-            summary = model_summary(self.learner)
-            try_mlflow_log(mlflow.set_tag, "model_summary", summary)
 
-            tempdir = tempfile.mkdtemp()
-            try:
-                summary_file = os.path.join(tempdir, "model_summary.txt")
-                with open(summary_file, "w") as f:
-                    f.write(summary)
-                try_mlflow_log(mlflow.log_artifact, local_path=summary_file)
-            finally:
-                shutil.rmtree(tempdir)
-
-        def on_train_end(self, **kwargs):
-            if log_models:
-                try_mlflow_log(log_model, self.learner, artifact_path="model")
+            def __init__(self, learner):
+                super().__init__(learner)
+                self.learner = learner
+                self.opt = self.learn.opt
+                self.metrics_names = ["train_loss", "valid_loss"] + [
+                    o.__name__ for o in learner.metrics
+                ]
+
+            def on_epoch_end(self, **kwargs):
+                """
+                Log loss and other metrics values after each epoch
+                """
+                if kwargs["smooth_loss"] is None or kwargs["last_metrics"] is None:
+                    return
+                epoch = kwargs["epoch"]
+                metrics = [kwargs["smooth_loss"]] + kwargs["last_metrics"]
+                metrics = map(float, metrics)
+                metrics = dict(zip(self.metrics_names, metrics))
+                metrics_logger.record_metrics(metrics, epoch)
+
+            def on_train_begin(self, **kwargs):
+                info = layers_info(self.learner)
+                try_mlflow_log(mlflow.log_param, "num_layers", len(info))
+                try_mlflow_log(mlflow.log_param, "opt_func", self.opt_func.func.__name__)
+
+                if hasattr(self.opt, "true_wd"):
+                    try_mlflow_log(mlflow.log_param, "true_wd", self.opt.true_wd)
+
+                if hasattr(self.opt, "bn_wd"):
+                    try_mlflow_log(mlflow.log_param, "bn_wd", self.opt.bn_wd)
+
+                if hasattr(self.opt, "train_bn"):
+                    try_mlflow_log(mlflow.log_param, "train_bn", self.train_bn)
+
+                summary = model_summary(self.learner)
+                try_mlflow_log(mlflow.set_tag, "model_summary", summary)
+
+                tempdir = tempfile.mkdtemp()
+                try:
+                    summary_file = os.path.join(tempdir, "model_summary.txt")
+                    with open(summary_file, "w") as f:
+                        f.write(summary)
+                    try_mlflow_log(mlflow.log_artifact, local_path=summary_file)
+                finally:
+                    shutil.rmtree(tempdir)
+
+            def on_train_end(self, **kwargs):
+                if log_models:
+                    try_mlflow_log(log_model, self.learner, artifact_path="model")
+
+        return __MLflowFastaiCallback
```

Comment:
Nit: Can we have these `get*Callback` methods across the various modules construct an instance of the callback class, or is there a case where we need to return the class definition instead?

Reply:
Yes we can, will update in commit

Reply:
Only pytorch we need to return class, because we check `isinstance(callbacks, __MLflowPLCallback)`.